service/matching/workers/registry_test.go

```diff
@@ -195,6 +195,112 @@ func TestRegistryImpl_ListWorkers(t *testing.T) {
 	}
 }
 
+// Exercises the query matching functionality of ListWorkers.
+func TestRegistryImpl_ListWorkersWithQuery(t *testing.T) {
+	tests := []struct {
+		name            string
+		setup           func(*registryImpl)
+		nsID            namespace.ID
+		query           string
+		expectedCount   int
+		expectedWorkers []string // WorkerInstanceKeys
+		expectError     bool
+	}{
+		{
+			name: "valid query - basic filtering",
+			setup: func(r *registryImpl) {
+				r.upsertHeartbeats("namespace1", []*workerpb.WorkerHeartbeat{
+					{WorkerInstanceKey: "worker1", TaskQueue: "queue1"},
+					{WorkerInstanceKey: "worker2", TaskQueue: "queue2"},
+				})
+			},
+			nsID:            "namespace1",
+			query:           "WorkerInstanceKey = 'worker1'",
+			expectedCount:   1,
+			expectedWorkers: []string{"worker1"},
+		},
+		{
+			name: "valid query - no matches",
+			setup: func(r *registryImpl) {
+				r.upsertHeartbeats("namespace1", []*workerpb.WorkerHeartbeat{
+					{WorkerInstanceKey: "worker1", TaskQueue: "queue1"},
+				})
+			},
+			nsID:            "namespace1",
+			query:           "TaskQueue = 'non-existent-queue'",
+			expectedCount:   0,
+			expectedWorkers: []string{},
+		},
+		{
+			name: "invalid query - malformed SQL",
+			setup: func(r *registryImpl) {
+				r.upsertHeartbeats("namespace1", []*workerpb.WorkerHeartbeat{
+					{WorkerInstanceKey: "worker1"},
+				})
+			},
+			nsID:        "namespace1",
+			query:       "invalid SQL syntax here",
+			expectError: true,
```

Comment:
it: would it be better to be able to provide an error substring that the error has to match?